### PR TITLE
[cross version tests] Clean up commands for installing dev versions

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

Clean up commands for installing dev versions.

## How is this patch tested?

Existing checks

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/ml-package-versions.yml [hunk 1]
```diff
@@ -2,8 +2,7 @@ sklearn:
   package_info:
     pip_release: "scikit-learn"
     install_dev: |
-      head_sha=$(git ls-remote https://github.com/scikit-learn/scikit-learn.git HEAD | cut -f1)
-      pip install git+https://github.com/scikit-learn/scikit-learn.git@$head_sha
+      pip install git+https://github.com/scikit-learn/scikit-learn.git
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:5]: I added `head_sha` to make it easier to identify the commit SHA but `pip freeze` shows it:

https://github.com/mlflow/mlflow/pull/4534/checks?check_run_id=2999114540#step:10:67

```
scikit-learn @ git+https://github.com/scikit-learn/scikit-learn.git@f2fb07581339f063b501b3e820917d66ee4c1fa3
```

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -242,12 +220,7 @@ onnx:
     pip_release: "onnx"
     install_dev: |
       sudo apt-get install protobuf-compiler libprotoc-dev
-      temp_dir=$(mktemp -d)
-      git clone https://github.com/onnx/onnx.git $temp_dir
-      git --git-dir=$temp_dir/.git rev-parse HEAD
-      cd $temp_dir
-      git submodule update --init --recursive
-      python setup.py install
+      pip install git+https://github.com/onnx/onnx.git
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:223]: `pip` automatically run `git submodule update --init --recursive` when the repo contains submodules.

File: mlflow/ml-package-versions.yml [hunk 3]
```diff
@@ -170,24 +160,12 @@ catboost:
...
-        cp $wheel_path $CACHE_DIR
-
-        # Install wheel
-        pip install $wheel_path
+      if [ ! -d "$CACHE_DIR" ] || [ -z $(find $CACHE_DIR -type f -name "catboost-*.whl") ]; then
+        head_sha=$(git ls-remote https://github.com/catboost/catboost.git HEAD | cut -f1)
+        pip wheel --no-deps --wheel-dir $CACHE_DIR \
+          git+https://github.com/catboost/catboost.git@$head_sha#subdirectory=catboost/python-package
       fi
+      pip install $(find $CACHE_DIR -type f -name "catboost-*.whl")
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:168]: For catboost, use `pip wheel` becuase we want to cache the built wheel.

**Comment by harupy** [comment]
The spacy and onnx failures seems unrelated to this PR.

**Review by WeichenXu123 — APPROVED** [review]
LGTM

**Comment by WeichenXu123** [comment]
I saw https://github.com/explosion/spaCy/pull/8559 and https://github.com/onnx/onnx-mlir/pull/225 was merged. Let's rerun the dev cross test again.

**Pull request merged**